ext/fetch/lib.rs

```diff
@@ -477,24 +477,41 @@ type BytesStream =
   Pin<Box<dyn Stream<Item = Result<bytes::Bytes, std::io::Error>> + Unpin>>;
 
 struct FetchResponseBodyResource {
-  reader: AsyncRefCell<StreamReader<BytesStream, bytes::Bytes>>,
+  reader: AsyncRefCell<Peekable<BytesStream>>,
   cancel: CancelHandle,
 }
 
+impl FetchResponseBodyResource {}
+
 impl Resource for FetchResponseBodyResource {
   fn name(&self) -> Cow<str> {
     "fetchResponseBody".into()
   }
 
-  fn read_return(
-    self: Rc<Self>,
-    mut buf: ZeroCopyBuf,
-  ) -> AsyncResult<(usize, ZeroCopyBuf)> {
+  fn read(self: Rc<Self>, limit: usize) -> AsyncResult<BufView> {
     Box::pin(async move {
-      let mut reader = RcRef::map(&self, |r| &r.reader).borrow_mut().await;
-      let cancel = RcRef::map(self, |r| &r.cancel);
-      let read = reader.read(&mut buf).try_or_cancel(cancel).await?;
-      Ok((read, buf))
+      let reader = RcRef::map(&self, |r| &r.reader).borrow_mut().await;
+
+      let fut = async move {
+        let mut reader = Pin::new(reader);
+        loop {
+          match reader.as_mut().peek_mut().await {
+            Some(Ok(chunk)) if !chunk.is_empty() => {
+              let len = min(limit, chunk.len());
+              let chunk = chunk.split_to(len);
+              break Ok(chunk.into());
+            }
+            Some(_) => match reader.as_mut().next().await.unwrap() {
```

Comment:
Is this unwrap safe here? If so, why?

Reply:
The logic is that `peek` returned `Some`, so this `.next()` must also return `Some`. This assumption is also made here: https://github.com/denoland/deno/blob/d13c88e70de9b3fd61c526ead90716fac503deed/ext/http/lib.rs#L362.

Because of the `.await`, I am not confident this is correct though. I'll see if I can switch this to `.try_next` (which must always succeed, because `.peek` returned `Some(_)`.

Reply:
Checked, and due to the implementation of `Peekable::poll_next`, it is safe & correct. Have added a comment.